### PR TITLE
Introduced upb_extreg and plumbed it into decoder.

### PR DESCRIPTION
**Pull request by haberman**
We also change `arena` to be the final parameter to `upb_decode()` as this follows a general pattern throughout upb APIs of making arena last.

### CODE REVIEW

File: cmake/google/protobuf/descriptor.upb.h [hunk 1]
```diff
@@ -162,13 +162,18 @@ UPB_INLINE google_protobuf_FileDescriptorSet *google_protobuf_FileDescriptorSet_
 UPB_INLINE google_protobuf_FileDescriptorSet *google_protobuf_FileDescriptorSet_parse(const char *buf, size_t size,
                         upb_arena *arena) {
   google_protobuf_FileDescriptorSet *ret = google_protobuf_FileDescriptorSet_new(arena);
-  return (ret && upb_decode(buf, size, ret, &google_protobuf_FileDescriptorSet_msginit, arena)) ? ret : NULL;
+  if (!ret || !upb_decode(buf, size, ret, &google_protobuf_FileDescriptorSet_msginit, arena)) return NULL;
```
[the remark this answers is not in the document]
Comment (haberman) [reply cmake/google/protobuf/descriptor.upb.h:165]: Done.

File: cmake/google/protobuf/descriptor.upb.h [hunk 2]
```diff
@@ -199,13 +204,18 @@ UPB_INLINE google_protobuf_FileDescriptorProto *google_protobuf_FileDescriptorPr
 UPB_INLINE google_protobuf_FileDescriptorProto *google_protobuf_FileDescriptorProto_parse(const char *buf, size_t size,
                         upb_arena *arena) {
   google_protobuf_FileDescriptorProto *ret = google_protobuf_FileDescriptorProto_new(arena);
-  return (ret && upb_decode(buf, size, ret, &google_protobuf_FileDescriptorProto_msginit, arena)) ? ret : NULL;
+  if (!ret || !upb_decode(buf, size, ret, &google_protobuf_FileDescriptorProto_msginit, arena)) return NULL;
```
[the remark this answers is not in the document]
Comment (haberman) [reply cmake/google/protobuf/descriptor.upb.h:207]: Done.

File: upb/msg.c [hunk 3]
```diff
@@ -338,3 +338,56 @@ bool _upb_mapsorter_pushmap(_upb_mapsorter *s, upb_descriptortype_t key_type,
...
+static void extreg_key(char *buf, const upb_msglayout *l, uint32_t fieldnum) {
+  memcpy(buf, &l, sizeof(l));
+  memcpy(buf + sizeof(l), &fieldnum, sizeof(fieldnum));
+}
+
+upb_extreg *upb_extreg_new(upb_arena *arena) {
+  upb_extreg *r = upb_arena_malloc(arena, sizeof(*r));
+  if (!r) return NULL;
+  r->arena = arena;
+  if (!upb_strtable_init(&r->exts, 8, arena)) return NULL;
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/msg.c:360]: The caller owns the arena, the memory will be freed when the caller frees the arena.

File: upb/msg.c [hunk 4]
```diff
@@ -338,3 +338,56 @@ bool _upb_mapsorter_pushmap(_upb_mapsorter *s, upb_descriptortype_t key_type,
...
+                             upb_value_constptr(e), r->arena)) {
+      /* Back out the entries previously added. */
+      for (end = e, e = start; e < end; e++) {
+        extreg_key(buf, e->extendee, e->field.number);
+        upb_strtable_remove(&r->exts, buf, EXTREG_KEY_SIZE, NULL);
+      }
+      return false;
+    }
+  }
+  return true;
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/msg.c:379]: Agreed, yes that reads nicer.

**Review by fowles — APPROVED** [review]

**Pull request merged**